linkerd/app/outbound/src/http/concrete.rs

```diff
@@ -1,104 +1,368 @@
                     Response = http::Response<http::BoxBody>,
-                    Error = ConcreteError,
+                    Error = Error,
                     Future = impl Send,
                 > + Clone,
         >,
     >
     where
-        N: svc::NewService<Endpoint, Service = NSvc> + Clone + Send + Sync + 'static,
+        // Concrete target type.
+        T: svc::Param<Dispatch>,
+        // T: svc::Param<svc::queue::Capacity>,
+        // T: svc::Param<svc::queue::Timeout>,
+        T: Clone + Debug + Send + Sync + 'static,
+        // Endpoint resolution.
+        R: Resolve<ConcreteAddr, Error = Error, Endpoint = Metadata>,
+        // Endpoint stack.
+        N: svc::NewService<Endpoint<T>, Service = NSvc> + Clone + Send + Sync + 'static,
         NSvc: svc::Service<http::Request<http::BoxBody>, Response = http::Response<http::BoxBody>>
             + Send
             + 'static,
         NSvc::Error: Into<Error>,
         NSvc::Future: Send,
-        R: Resolve<Concrete, Error = Error, Endpoint = Metadata>,
     {
-        self.map_stack(|config, rt, endpoint| {
-            endpoint
+        let resolve =
+            svc::MapTargetLayer::new(|t: Balance<T>| -> ConcreteAddr { ConcreteAddr(t.addr) })
+                .layer(resolve.into_service());
+
+        self.map_stack(|config, rt, inner| {
+            let inbound_ips = config.inbound_ips.clone();
+
+            let forward = inner
+                .clone()
+                .push_on_service(
+                    rt.metrics
+                        .proxy
+                        .stack
+                        .layer(stack_labels("http", "forward")),
+                )
+                .instrument(|e: &Endpoint<T>| info_span!("forward", addr = %e.addr));
+
+            let endpoint = inner
                 .push_on_service(
                     rt.metrics
                         .proxy
                         .stack
                         .layer(stack_labels("http", "endpoint")),
                 )
-                .instrument(|e: &Endpoint| info_span!("endpoint", addr = %e.addr))
-                .lift_new()
-                .push(endpoint::NewFromMetadata::layer(config.inbound_ips.clone()))
+                .instrument(|e: &Endpoint<T>| info_span!("endpoint", addr = %e.addr));
+
+            let inbound_ips_ = inbound_ips.clone();
+            let balance = endpoint
+                .push_map_target(
+                    move |((addr, metadata), target): ((SocketAddr, Metadata), Balance<T>)| {
+                        tracing::trace!(%addr, ?metadata, ?target, "Resolved endpoint");
+                        let is_local = inbound_ips_.contains(&addr.ip());
+                        Endpoint {
+                            addr: Remote(ServerAddr(addr)),
+                            metadata,
+                            is_local,
+                            parent: target.parent,
+                        }
+                    },
+                )
```

Comment:
silly style nit, not important: we could avoid having to write `inbound_ips_` by writing this instead:
```suggestion
            let balance = endpoint
                .push_map_target({
                    let inbound_ips = inbound_ips.clone();
                    move |((addr, metadata), target): ((SocketAddr, Metadata), Balance<T>)| {
                        tracing::trace!(%addr, ?metadata, ?target, "Resolved endpoint");
                        let is_local = inbound_ips.contains(&addr.ip());
                        Endpoint {
                            addr: Remote(ServerAddr(addr)),
                            metadata,
                            is_local,
                            parent: target.parent,
                        }
                    }
                })
```
up to you whether that's nicer, though...